src/transformers/models/modernbert_decoder/modular_modernbert_decoder.py

```diff
@@ -0,0 +1,968 @@
+
+        # Transpose to [batch_size, num_heads, seq_len, head_dim]
+        query = query.transpose(1, 2)
+        key = key.transpose(1, 2)
+        value = value.transpose(1, 2)
+
+        # Apply rotary embeddings (passed from model level)
+        cos, sin = position_embeddings
+        query, key = apply_rotary_pos_emb(query, key, cos, sin)
+
+        if past_key_value is not None:
+            # sin and cos are specific to RoPE models; cache_position needed for the static cache
+            cache_kwargs = {"sin": sin, "cos": cos, "cache_position": cache_position}
+            key, value = past_key_value.update(key, value, self.layer_idx, cache_kwargs)
+
+        attention_interface = eager_attention_forward
+        if self.config._attn_implementation != "eager":
+            attention_interface = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
+
+        # Pass sliding window parameter for sliding attention layers
+        sliding_window_param = self.local_attention[0] if self.local_attention[0] != -1 else None
+
+        attn_outputs = attention_interface(
+            self,
+            query,
+            key,
+            value,
+            attention_mask,
+            dropout=0.0 if not self.training else self.attention_dropout,
+            sliding_window=sliding_window_param,
+            is_causal=True,
+            **kwargs,
+        )
+
+        attn_output = attn_outputs[0]
+        attn_weights = attn_outputs[1] if output_attentions and len(attn_outputs) > 1 else None
+
+        # Reshape to [batch_size, seq_len, hidden_size] - this handles both eager and FA2 outputs
+        input_shape = hidden_states.shape[:-1]
+        attn_output = attn_output.reshape(*input_shape, -1).contiguous()
+
+        # Apply output projection
+        hidden_states = self.out_drop(self.Wo(attn_output))
+
+        outputs = (hidden_states,)
+        if output_attentions:
+            outputs += (attn_weights,)
+        if past_key_value is not None:
+            outputs += (past_key_value,)
+
+        return outputs
+
+
+class ModernBertDecoderLayer(GradientCheckpointingLayer):
+    def __init__(self, config: ModernBertDecoderConfig, layer_idx: Optional[int] = None):
+        super().__init__()
+        self.config = config
+        self.layer_idx = layer_idx
+        self.attention_type = config.layer_types[layer_idx]
+
+        if layer_idx == 0:
+            self.attn_norm = nn.Identity()
+        else:
+            self.attn_norm = nn.LayerNorm(config.hidden_size, eps=config.norm_eps, bias=config.norm_bias)
+        self.attn = ModernBertDecoderAttention(config=config, layer_idx=layer_idx)
+        self.mlp_norm = nn.LayerNorm(config.hidden_size, eps=config.norm_eps, bias=config.norm_bias)
+        self.mlp = ModernBertMLP(config)
+
+    @torch.compile(dynamic=True)
+    def compiled_mlp(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        return self.mlp(self.mlp_norm(hidden_states))
```

Comment:
I don't think this is the way we usually go, in general we recommend using model.forward's compile and using static cache, or something like that. Compile is not supported on mps for example (not well supported)